feat: add Esri ArcGIS Maps SDK adapter

Tugark opened the pull request:
Hope you got back well from FOSS4G! ;)

As discussed at FOSS4G, I tried to whip up and Esri ArcGIS Maps SDK adapter.

The functionality is working, so if you want to glance over it, feel free to add suggestions. Couple of things to note:

* I am using a single [GraphicsLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-GraphicsLayer.html) instead of a [GeoJSONLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-GeoJSONLayer.html). The advantage being that it's less abstraction and we don't have to maintain several layers and split the features accordingly (since the GeoJSONLayer only accepts features of the same type along [some other limitations](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-GeoJSONLayer.html#limits)).
  * However, this could be fixed with some more logic, should we require the more powerful GeoJSONLayer (or, [FeatureLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-FeatureLayer.html) at that)
* The drawing style is actually not consistent with other libraries in that the width is calculated differently - lines are bolder, points are smaller). I'm not sure if it's your goal to have the "look" identical across libraries; if so, I'd have to dig deeper into the drawing logic of the Esri tool.

I'm opening this PR as a draft because it is still missing tests - I have to work on them as I get the nice old `SyntaxError: Unexpected token export` and the Esri API is a bit cumbersome to test or mock. I hope I get around it somewhen soon, but I felt that I could already get some feedback on how that works for you.

Once again, thanks for your work so far. I do believe that this library has a lot of potential!

JamesLMilner commented:
Hey @Tugark  - I made it back in one piece thank you. It was great meeting you at FOSS4G this year, thanks for coming to the talk!

This is awesome :tada: Thanks so much for raising it. I am just on holiday till Monday but will take a look properly when I am back.

I don't think it's necessary to have 100% identical styling, I feel 'best effort' is acceptable as long as they look approximately the same. It can always be a follow up improvement if it proves to be problematic.

JamesLMilner commented on src/adapters/arcgis-maps-sdk.adapter.ts:
```diff
@@ -0,0 +1,247 @@
+		let symbol: Symbol | undefined = undefined; // eslint-disable-line @typescript-eslint/ban-types
+		let geometry: Geometry | undefined = undefined;
+
+		switch (type) {
+			case "Point":
+				geometry = new this._lib.Point({
+					latitude: coordinates[1],
+					longitude: coordinates[0],
+				});
+				symbol = new this._lib.SimpleMarkerSymbol({
+					color: this.getColorFromHex(style.pointColor),
+					size: style.pointWidth,
+					outline: {
+						color: this.getColorFromHex(style.pointOutlineColor),
+						width: style.pointOutlineWidth,
+					},
+				});
+				break;
+			case "LineString":
+				geometry = new this._lib.Polyline({ paths: [coordinates] });
+				symbol = new this._lib.SimpleLineSymbol({
+					color: this.getColorFromHex(style.lineStringColor),
+					width: style.lineStringWidth,
+				});
+				break;
+			case "Polygon":
+				geometry = new this._lib.Polygon({ rings: coordinates });
+				symbol = new this._lib.SimpleFillSymbol({
+					color: this.getColorFromHex(
+						style.polygonFillColor,
+						style.polygonFillOpacity
+					),
+					outline: {
+						color: this.getColorFromHex(style.polygonOutlineColor),
+						width: style.polygonOutlineWidth,
+					},
+				});
+				break;
+		}
```
The following changes will bring the styling in line with the other adapters :) 

```typescript
		switch (type) {
			case "Point":
				geometry = new this._lib.Point({
					latitude: coordinates[1],
					longitude: coordinates[0],
				});
				symbol = new this._lib.SimpleMarkerSymbol({
					color: this.getColorFromHex(style.pointColor),
					size: style.pointWidth * 2 + "px",
					outline: {
						color: this.getColorFromHex(style.pointOutlineColor),
						width: style.pointOutlineWidth + "px",
					},
				});
				break;
			case "LineString":
				geometry = new this._lib.Polyline({ paths: [coordinates] });
				symbol = new this._lib.SimpleLineSymbol({
					color: this.getColorFromHex(style.lineStringColor),
					width: style.lineStringWidth + "px",
				});
				break;
			case "Polygon":
				geometry = new this._lib.Polygon({ rings: coordinates });
				symbol = new this._lib.SimpleFillSymbol({
					color: this.getColorFromHex(
						style.polygonFillColor,
						style.polygonFillOpacity
					),
					outline: {
						color: this.getColorFromHex(style.polygonOutlineColor),
						width: style.polygonOutlineWidth + "px",
					},
				});
				break;
```

Tugark replied:
Great, I've added that.

JamesLMilner commented:
Just checked this out - nice work! Code quality is great and all working nicely. I played with the styling and figured out you just need to set any widths to pixels and it matches the styling of the other adpaters. I also needed to set the point size to * 2 (assuming this is the radius in other adapters hence having to double it to get the width).

If you manage to get a chance to write some tests that would be great - current threshold is set to 65% so if we could try and hit that it would be awesome. Let me know if it becomes to arduous and we can look at getting it in without and I can try and follow it up when i get a moment (There is already an open issue for better adapter testing #18)

Tugark commented:
Hey @JamesLMilner Thanks for the review! I just added your suggestion and I also added some basic tests. I'll try to add some tests for the `render` method and the map initializer as well so the test coverage should be similar to the leaflet adapter tests.

I'm currently quite swamped, unfortunately, but I hope I'll be able to squeeze it in after the 23rd.

JamesLMilner commented:
@Tugark awesome, let me know if I can help at all. No worries on the timings, whenever is convenient for you. It's looking great, excited to get it in!

Tugark commented:
@JamesLMilner Finally got around for adding some more tests - as always, it's a bit messy because it's relatively cumbersome to properly mock the javascript API.

According to the coverage, it looks well:
![image](https://github.com/JamesLMilner/terra-draw/assets/16555423/e423537f-6584-4016-8dc8-ed8914eb1e75)

The uncovered lines are callbacks which I'm not sure can be tested. The rest should be good enough for the moment.

Let me know if you want me to clean up the tests even more.

JamesLMilner commented:
@Tugark amazing work! Thanks so much - I think this is ready to go in now :)
